### PR TITLE
Allow extending xast children using module augmentation

### PR DESCRIPTION
**Pull request by remcohaszing**
This is an idea based on syntax-tree/xastscript#8.

The idea is to use an approach similar to [`HTMLElementTagNameMap`](https://github.com/microsoft/TypeScript/blob/master/lib/lib.dom.d.ts#L19398-L19518). Each parent node is accompanied by an interface named `<name>ChildMap`. Only elements registered in this map may be used as children for the parent node they belong to. The jsdocs contain examples of how to register custom nodes. The same principle could be applied to other unist packages.

This could be taken a step further by making `unist#Parent` generic.

```ts
// unist
export interface Parent<ChildMap = Record<string, Node>> extends Node {
  children: Array<ChildMap[keyof ChildMap]>;
}

// xast
export interface Root extends UnistParent<RootChildMap> {
    type: 'root';
}
```

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: syntax-tree/xastscript#8
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by ChristianMurphy — APPROVED** [review]
Thanks @remcohaszing!

**Review by wooorm — APPROVED** [review]
Looks cool to me!

File: types/xast/index.d.ts [hunk 1]
```diff
@@ -12,18 +12,36 @@ import { Parent as UnistParent, Literal as UnistLiteral, Node as UnistNode } fro
 export { UnistNode as Node };
 
 /**
- * Node in xast containing other nodes.
- * Its content is limited to only other xast content.
+ * Node in xast containing a value.
  */
-export interface Parent extends UnistParent {
```
Comment (remcohaszing) [inline types/xast/index.d.ts:18]: I think I’m going to have to turn this into a `unist.Parent` re-export to prevent breaking changes.

**Comment by remcohaszing** [comment]
What do you think of the idea of making `unist.Parent` generic?

Also if this is going to be accepted, I’m going to apply the same on `hast` and `mdast`.

**Comment by wooorm** [comment]
You mean a similar child map for unist? What would the benefit there be?

**Review by wooorm — APPROVED** [review]

**Review by ChristianMurphy — APPROVED** [review]

**Comment by remcohaszing** [comment]
Ready to merge

**Pull request merged**